src/bin/josh_sync.rs

```diff
@@ -0,0 +1,169 @@
+                        &result.merge_commit_message,
+                    )? {
+                        println!(
+                            "Now push the current branch to {} (either a fork or the main repo) and create a PR",
+                            ctx.config.repo
+                        );
+                    }
+                }
+                Err(RustcPullError::NothingToPull) => {
+                    eprintln!("Nothing to pull");
+                    std::process::exit(2);
+                }
+                Err(RustcPullError::PullFailed(error)) => {
+                    eprintln!("Pull failure: {error:?}");
+                    std::process::exit(1);
+                }
+            }
+        }
+        Command::Push {
+            username,
+            branch,
+            config_path,
+            rust_version_path,
+        } => {
+            let ctx = load_context(&config_path, &rust_version_path)?;
+            let josh = get_josh_proxy()?;
+            let sync = GitSync::new(ctx.clone(), josh);
+            sync.rustc_push(&username, &branch)
+                .context("cannot perform push")?;
+
+            // Open PR with `subtree update` title to silence the `no-merges` triagebot check
+            println!(
+                r#"You can create the rustc PR using the following URL:
+https://github.com/{UPSTREAM_REPO}/compare/{username}:{branch}?quick_pull=1&title={}+subtree+update&body=r?+@ghost"#,
+                ctx.config.repo
+            );
+        }
+    }
+
+    Ok(())
+}
+
+fn load_context(config_path: &Path, rust_version_path: &Path) -> anyhow::Result<SyncContext> {
+    let config = load_config(&config_path)
+        .context("cannot load config. Run the `init` command to initialize it.")?;
+    let rust_version = std::fs::read_to_string(&rust_version_path)
+        .inspect_err(|err| eprintln!("Cannot load rust-version file: {err:?}"))
+        .map(Some)
+        .unwrap_or_default();
+    Ok(SyncContext {
+        config,
+        last_upstream_sha_path: rust_version_path.to_path_buf(),
+        last_upstream_sha: rust_version,
+    })
+}
+
+fn maybe_create_gh_pr(repo: &str, title: &str, description: &str) -> anyhow::Result<bool> {
+    if which::which("gh").is_ok()
+        && prompt("Do you want to create a rustc pull PR using the `gh` tool?")
+    {
+        std::process::Command::new("gh")
+            .args(&[
+                "pr",
+                "create",
+                "--title",
+                title,
+                "--body",
+                description,
+                "--repo",
+                repo,
+            ])
+            .spawn()?
+            .wait()?;
+        Ok(true)
+    } else {
+        Ok(false)
+    }
+}
+
+fn get_josh_proxy() -> anyhow::Result<JoshProxy> {
+    match JoshProxy::lookup() {
+        Some(proxy) => Ok(proxy),
+        None => {
+            if prompt("josh-proxy not found. Do you want to install it?") {
+                match try_install_josh() {
```

Comment:
Is there any way to detect an outdated josh and also run the install command then?

Reply:
It does not seem to have any `--version` flag :(

Reply:
Testing locally it looks like ` josh-filter -V` works but `josh-proxy` has no such thing? Not sure if we can rely on them being coupled

Reply:
```
$ cargo +stable install josh-proxy --git https://github.com/josh-project/josh --tag r24.10.04
    Updating git repository `https://github.com/josh-project/josh`
     Ignored package `josh-proxy v22.4.15 (https://github.com/josh-project/josh?tag=r24.10.04#61643c03)` is already installed, use --force to override
```
Cargo already tracks that for us. So it seems if we run the command unconditionally it will update if the installed version is out-of-date.

Reply:
Not sure if it's best practice to just unconditionally update an installed package, but I guess in this case it shouldn't be a big deal. I suppose it's much better than using old josh by accident.

Reply:
I guess we could grab the version output from Cargo's list?

```text
$ cargo install --list | grep josh-proxy
josh-proxy v22.4.15 (https://github.com/josh-project/josh?tag=r24.10.04#61643c03):
    josh-proxy
```

Reply:
Sadly, josh doesn't update its crate version. `v22.4.15` has been used for a while. We would have to parse the `tag=` part of the output.. I don't think that's worth it.